### ISSUE TITLE
[Feature]: Add embedding model functionality to tpu-inference

### ISSUE THREAD

**Issue by bvrockwell** [vllm-project/tpu-inference#899]
### 🚀 The feature, motivation and pitch

copied from [feature request in vllm upstream](https://github.com/vllm-project/vllm/issues/20869)

Great first issue!

Should enable for both torchax and jax (if needed).

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://github.com/vllm-project/tpu-inference/tree/main/docs), which can answer lots of frequently asked questions.

**Comment by carlesoctav** [comment]
I’ve already done some preliminary coding and research on how we might achieve it, but I’m still not entirely sure what the best approach is. Could someone review my plan and give feedback?
For now, I’ll try integrating the Torchax version first:

I’ll add a new metadata class similar to `TPUSamplingMetadata`, called `TPUPoolingMetadata`, and construct it in `runner._prepare_input()`. This metadata will handle pooling, analogous to how sampling works in `layers/jax/sample`. Initially, I thought we could follow the same pattern as the sampler, creating a pooling function that takes `hidden_states` and `TPUPoolingMetadata` as arguments (see next paragraph for the issue).

The problem is that when using `vllm_model_wrapper`, pooling is actually part of the Torch model (under `model.pooler`). In contrast, sampling doesn’t depend on the Torch model, it just uses logits and sampling metadata. The pooler requires both `hidden_states` and the (Torch) pooling metadata. Because of this model dependency, should we instead add a new `jit_pool_fn` inside `VLLMModelWrapper` rather than mirroring how `sample` works (which is implemented purely in JAX even though we use Torchax)?

```python
    def jit_pool_fn(self):
        @jax.jit
        def pool_fn(
            self,
            params_and_buffers: Any,
            hidden_states: jax.Array,
            pooling_metadata: TPUPoolingMetadata,
            lora_metadata,
        ) -> list[torch.Tensor | None]:
            if not self._is_pooling_model:
                raise RuntimeError("Pooling requested for a non-pooling model")

            hidden_states_torch = torch_view(hidden_states)
            lora_metadata = torch_view(lora_metadata)

            with torchax.default_env(), set_vllm_model_wrapper_context(
                    kv_caches=None, mesh=self.mesh):
                original_lora_metadata = replace_lora_metadata(
                    self.model, lora_metadata, self.vllm_config.lora_config)

                # Not sure if this will work; the original pool takes the TPUPooling metadata, which was a pytree of arrays,
                # and we convert it to the torch version without doing torch_view?

                vllm_pooling_metadata = pooling_metadata.to_vllm_metadata()
                # @dataclass
                # class PoolingCursor:
                #     index: list[int]
                #     first_token_indices_gpu: torch.Tensor
                #     last_token_indices_gpu: torch.Tensor
                #     prompt_lens_cpu: torch.Tensor
                #     num_scheduled_tokens_cpu: torch.Tensor
                #
                #     def __getitem__(self, indices: slice):
                #         return PoolingCursor(
                #             index=self.index[indices],
                #             first_token_indices_gpu=self.first_token_indices_gpu[indices],
                #             last_token_indices_gpu=self.last_token_indices_gpu[indices],
                #             prompt_lens_cpu=self.prompt_lens_cpu[indices],
                #             num_scheduled_tokens_cpu=self.num_scheduled_tokens_cpu[indices],
                #         )
                #
                #     def is_partial_prefill(self):
                #         return not torch.all(self.prompt_lens_cpu == self.num_scheduled_tokens_cpu)
                vllm_pooling_metadata.build_pooling_cursor(
                    pooling_metadata.scheduled_tokens_list(),
                    hidden_states_torch.device,
                )

                pool_output = torch.func.functional_call(
                    self.model.vllm_model.pooler,
                    torch_view(params_and_buffers),
                    kwargs={
                        "pooling_hidden_states": hidden_states_torch,
                        "pooling_metadata": vllm_pooling_metadata,
                    },
                    tie_weights=False,
                )
                return jax_view(pool_output)
        return pool_fn

```
Alternatively, we could keep pooling entirely in JAX (similar to how sampling works), but that would require work to determine when and how to create the `nnx` pooling module (or maybe keep it simple, just a function). We’d likely need this approach anyway if we plan to add full JAX-based pooling capabilities later. (for the sake of parallelism, on torch, sampling was done on `Sampler` class which a `nn.Module` but we do sampling on jax.)

Any thoughts?
Would you prefer to keep pooling purely in JAX (by creating a new `nnx` module), or is it acceptable to patch this for Torchax support for now?

**Comment by carlesoctav** [comment]
For nnx model_impl, you might run into a path/key mismatch issue when trying to load the model for now, on the first layer (embed_token). I’ve tested this on Qwen/Qwen3-Embedding-0.6B.

**Comment by carlesoctav** [comment]
1. For the torchax version, keep the loading process as is: it will use `get_vllm_model` and create the model_class. If we use a task that is not CausalLM/generation, it will actually run the adapter (see `vllm/model_executor/models/adapters.py`) and convert the model for the specific task (e.g., `ModelForEmbedding`, etc.). We just need to detect if a certain task is being used and create a `PoolerFor{Task}` that only performs the computations relevant to this task (it will takes `hidden_states`, `jax_pooling_metadata`). This will be a pure Flax/nnx implementation. We'll store this adapter as pooler on the `tpu_model_runner`. Notes that, the model weights for this adapter will be loaded in both JAX and Torch; we need a post-processing step to maybe remove the model_weight of the torch-adapter (but for embedding task, there's no model weights)

2. If we use the nnx implementation, do the same as `get_vllm_model`: convert the model for the task (note that it will have a different way to load the model, like removing prefix or suffix from the safe_tensor path). Finally, We just need to link `model.pooler` to `runner.pooler`.

3. In `tpu_jax_model_runner`, we need to store both the `is_pooling_model` flag and the pooler. In `_prepare_inputs()`, create a `TPUSupportedPoolingMetadata` class (on `jax/layers/pool/pooling_metadata.py`). In `_execute_model`, add a branch to check if pooling needs to be run by checking `self.is_pooling_model`. If so, run `pool(hidden_states, pooling_metadata, pooler: nnx.Module)` (`jax/layers/pool/pooling.py`). We also need to branch when compiling this pool function: if `is_pooling_model is True`, there is no need to run functions related to generation (`compute_logits`, `sampling`, etc.).

**Comment by bvrockwell** [comment]
cc @vkantamani-cienet

**Comment by bvrockwell** [comment]
Hey @carlesoctav if you take a stab at doing this in a way that torchax and jax models are supported, we can assign a reviewer.

**Comment by carlesoctav** [comment]
Hi, sure, i'd love to take on this issue.
